### PR TITLE
Disable unauthorized buttons

### PR DESCRIPTION
**Pull request by datamel**
Putting this up as an initial PR to disable buttons for mutations that are not authorized.
I have not dealt with the tooltip, the `"workflow-play-pause-button"` or `"workflow-stop-button"` yet, will do a follow-up pr for those. 

This branch requires it's sibling PR: https://github.com/cylc/cylc-uiserver/pull/204
These changes partially address https://github.com/cylc/cylc-ui/issues/704

Changes look as follows: 
For `userB`, accessing `userA`'s ui-server with the following auth permissions in UserA's config files:

```
c.UIServer.site_authorization = {
    "*": {
        "*": {
            "default": ["ALL"],
            "limit": ["ALL"],}
    }
}

c.UIServer.user_authorization = {
    'userB': ['READ', 'trigger']
}
```

Going to : `/userA/userprofile`, when authenticated as `userB` will look as follows:

image

And, clicking on the mutation button of a workflow will look as follows:

image

Will put this up as a Draft PR for now... I am going to have a look at, rather than removing these buttons, greying these buttons out.

Update: Greyed out buttons now functioning and tested.

- [x] I have read `CONTRIBUTING.md` and added my name as a Code Contributor.
- [x] Contains logically grouped changes (else tidy your branch by rebase).
- [x] Does not contain off-topic changes (use other PRs for other changes).
- [x] Appropriate tests are included (unit and/or functional).
- [x] Appropriate change log entry included.
- [x] No documentation update required (Will be covered by documentation for authorization).

### CODE REVIEW

**Comment by datamel** [comment]
Update: Buttons are now greyed out:

image

Will investigate testing this and getting the ci working next.

File: src/components/cylc/cylcObject/Menu.vue
Comment (kinow) [inline src/components/cylc/cylcObject/Menu.vue:166]: See comments above; I think here every time `this.mutations` or `this.user` changes, this `displayMutations` computed property will be re-evaluated. In this part, the `this.user.permissions.map` will be executed for each mutation. Using other computer property instead, I think it could be simpler to read the relationship between reactive properties in Vue, and would avoid re-calculating the `this.user.permissions.map((str) => str.toLowerCase())`.

But I don't trust my code after 4-5 PM and without :coffee: , and the PR is still in draft... but just in case you'd like to have fun experimenting more with the Vue reactivity system @datamel :tada:

**Review by kinow — COMMENTED** [review]
Amazing work @datamel !!! Added a few comments, but nothing major. Happy to review it if you'd like me to, once you are ready :+1:

**Comment by kinow** [comment]
>Will investigate testing this and getting the ci working next.

Huh, e2e tests... here's what I do when that happens:

- view the github actions output (either that fancy output, or raw logs if the former is taking too long to load)
- find which tests are failing
- start e2e tests with `yarn run test:e2e`
- pick the same browser that failed in the GH actions (the name of the job should include a browser name...)
- execute the tests and reproduce the error locally

Then you leave the `yarn run test:e2e` running, it will be watching for changes and executing the tests as you change the code. If you know which test is failing, one think I like doing is renaming it so that Cypress runs just that test. e.g.

```js
describe('Test 1, () => {})
....
describe.only('Test 15', () => {})
```

If you append `.only` to a `describe` or `it`, then as you change the code Cypress will re-run only that test. That should save you some time. Then all you gotta do is just figure out what's wrong in the Cypress test, fix it, commit, push :+1:

**Comment by datamel** [comment]
Thanks so much for the tips @kinow. I did a Vue course last Friday and thought I would give this one a go, very grateful for the comments!

**Comment by hjoliver** [comment]
NIce, thanks @datamel :medal_military:

**Comment by datamel** [comment]
Marking this as ready for review. Its sibling https://github.com/cylc/cylc-uiserver/pull/204 is also ready for final review. 
I have tried to implement your suggested changes @kinow, and have added tests.  
I have experienced a problem, which I am unsure if it will carry through to production which is the page sometimes needs hard refreshing before the mutations are greyed out, perhaps this is just a caching issue for me. I shall look into this further if it is a persistent problem.

**Review by kinow — COMMENTED** [review]
Great work @datamel ! Minor comments, but no blockers :-)

Checked out the Cylc UIS https://github.com/cylc/cylc-uiserver/pull/204 PR to test this PR. @datamel I later realized I had invalid configuration key, but:

image

image

Then I removed my user from the config file (i.e removed the line with `"kinow": ["ALL"]`) and restarted the UIS. @datamel I got the same list of permissions I think, but in different order? :confused: 

image

Then tested my own user accessing `kinow|five`:

image

And tried to access it now with `cylc` with this corrected configuration:

```
c.CylcUIServer.user_authorization = {
  "*": ["READ"],            # any authenticated user has READ access
  "cylc": [
    "ALL", "!play", "!stop"
  ]                         # cylc has ALL operations, but not play and stop
}
```

image

@datamel I think at the moment the app is failing to display the UI, but it might be due to something in the UIS rejecting the request. Maybe we should display a simple message instead of a blank page, so users are aware of what happened. Not related to your PR IMO since this is about unauthorized buttons :)

Anyway, I can't get the UI to display the `kinow|five` workflow, even though I have given permission to the `cylc` user? From the logs:

```
...
[D 2021-09-22 12:20:49.929 CylcUIServer] Looking for jupyter_cylc_config in /home/kinow/.cylc/hub
[D 2021-09-22 12:20:49.929 CylcUIServer] Looking for jupyter_cylc_config in /etc/cylc
[D 2021-09-22 12:20:49.929 CylcUIServer] Looking for jupyter_cylc_config in /home/kinow/Development/python/workspace/cylc-uiserver/cylc/uiserver
[I 2021-09-22 12:20:49.936 CylcHubApp manager:345] cylc.uiserver | extension was successfully linked.
[D 2021-09-22 12:20:49.936 CylcHubApp serverapp:1881] Raising open file limit: soft 1024->4096; hard 1048576->1048576
[I 2021-09-22 12:20:50.001 CylcUIServer] Starting Cylc UI Server
[I 2021-09-22 12:20:50.001 CylcUIServer] Serving UI from: /home/kinow/Development/python/workspace/cylc-ui/dist
[I 2021-09-22 12:20:50.004 CylcHubApp manager:367] cylc.uiserver | extension was successfully loaded.
[I 2021-09-22 12:20:50.004 CylcHubApp mixins:576] Starting jupyterhub-singleuser server version 1.4.2
...
[I 2021-09-22 12:16:11.489 CylcHubApp log:189] 101 GET /user/kinow/cylc/subscriptions (cylc@::1) 1.05ms
[W 2021-09-22 12:16:11.814 CylcHubApp handlers:128] Authorisation failed for cylc
[D 2021-09-22 12:16:11.814 CylcHubApp handlers:253] Using contents: services/contents
[D 2021-09-22 12:16:11.814 CylcHubApp handlers:253] Using contents: services/contents
[W 2021-09-22 12:16:11.816 CylcHubApp log:189] 403 GET /user/kinow/cylc/userprofile (cylc@::1) 2.31ms
WARNING:cylc.uiserver.authorise:Authorization failed for cylc:requested to read.
[I 2021-09-22 12:16:11.849 CylcHubApp log:189] 200 POST /user/kinow/cylc/graphql (cylc@::1) 32.12ms
[D 2021-09-22 12:16:51.267 CylcHubApp mixins:518] Notifying Hub of activity 2021-09-22T00:12:19.293007Z

```

@datamel I wonder if we should log the permissions loaded too, maybe in debug/trace level?

**Comment by kinow** [comment]
And great job on getting the tests working @datamel ! Normally the lengthiest task in UI PR's :rofl:

**Comment by hjoliver** [comment]
(Is this good to go now @kinow ?)

**Comment by datamel** [comment]
> > (Is this good to go now @kinow ?)
> 
> I haven't tested it, but I saw changes to its sibling PR in cylc-uiserver this week, some days ago. @datamel is it ready for review, or are you still working on the flow & uis PRs?

Yes, ready for review. I had pushed some changes to the [uis pr](https://github.com/cylc/cylc-uiserver/pull/204) after chatting to Oliver - this has changed the way the authorised user is passed to the middleware, it is now passed through the context on the info object. This is preferable to the original way of overwriting the user on on the subscription server object. This should not result in any functionality change but is a cleaner way of doing it, and a bit safer.

**Comment by AaronDCole** [comment]
> Not a problem @datamel .
> 
> I'm leaving my +1 here as your change is working fine right now, contains test, and the code looks good too!
> 
> Later, when you have rebased and updated the code for the the play/stop buttons, @AaronDCole can complete the review. If you have finished before Wednesday/UK next week I can do a quick sanity check before Aaron too.
> 
> Thanks!!! 👏 🏆

This looks good, but I'll wait for Bruno to step me through his review process before I merge it.

**Comment by hjoliver** [comment]
The back-end auth PR is now merged (and I was testing it with this cylc-ui branch).

**Review by hjoliver — APPROVED** [review]
Looks good, tests as working nicely.  

@datamel - looks like there's one or two very minor comments from @oliver-sanders that you could consider addressing as follow-up?

**Comment by hjoliver** [comment]
More than enough approvals :tada:

**Pull request merged**
